File: BahamasEngine/BahamasEngine/MetaDataManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace BahamasEngine
{
    public sealed class MetaDataManager
    {
        public static Dictionary<string, OptionContract> OptionContracts { get; private set; }
        public static Dictionary<string, OptionChainSnapshot>[] OptionChainHistory { get; private set; }
        public static Dictionary<string, OptionChain> OptionChains { get; private set; }
        public static string[] TradingDates { get; private set; }

        private MetaDataManager(){}

        public static void ImportMetaData(string ticker)
        {
            OptionContracts = new Dictionary<string, OptionContract>();
            OptionChains = new Dictionary<string, OptionChain>();

            Console.WriteLine("Starting data load...");
            TradingDates = File.ReadAllLines(Settings.DataPath + $@"{ticker}\" + "TRADINGDATES.txt");

            Console.WriteLine("     Loading Contract MetaData");
            ImportOptionMetaData(ticker);

            Console.WriteLine("     Loading data complete.");
        }

        private static void ImportOptionMetaData(string ticker)
        {
            OptionChainHistory = new Dictionary<string, OptionChainSnapshot>[TradingDates.Length];
            for (int dateIndex = 0; dateIndex < TradingDates.Length; dateIndex++)
            {
                OptionChainHistory[dateIndex] = new Dictionary<string, OptionChainSnapshot>();

                string date = TradingDates[dateIndex];
                var manifestData = File.ReadAllText(Settings.DataPath + $@"{ticker}\" + "MANIFEST_" + date + ".csv").Split('\n');

                foreach (var line in manifestData)
                {
                    string[] rowData = line.Split(',');
                    if (rowData.Length < 5)
                        continue;

                    char type = rowData[4][0];
                    double strike = Convert.ToDouble(rowData[5]);
                    string chainId = rowData[0];
                    string optionId = rowData[1];

                    OptionChain chain = new OptionChain(chainId, Settings.Ticker, rowData[2], rowData[3], 100);
                    OptionChainSnapshot chainSnapshot = new OptionChainSnapshot(chain);
                    OptionContract contract = new OptionContract(optionId, Settings.Ticker, rowData[2], rowData[3], strike, type, chainId);

                    if (!OptionChains.ContainsKey(chainId))
                        OptionChains.Add(chainId, chain);

                    if (!OptionContracts.ContainsKey(contract.ID))
                        OptionContracts.Add(contract.ID, contract);

                    if (!OptionChainHistory[dateIndex].ContainsKey(chainId))
                        OptionChainHistory[dateIndex].Add(chainId, chainSnapshot);

                    OptionChainSnapshot targetchain = OptionChainHistory[dateIndex][chainId];

                    if (type == 'C')
                    {
                        targetchain.CallOptionContracts.Add(contract);
                    }
                    else if (type == 'P')
                    {
                        targetchain.PutOptionContracts.Add(contract);
                    }
                    else
                    {
                        throw new InvalidDataException();
                    }
                }
            }
        }
    }
}

File: BahamasEngine/metadata_manager.py
import csv

import settings
import option_chain
import option_contract

trading_dates = []
option_contracts = {}
option_chains = {}

def import_metadata():
    print("Loading metadata...")

    global trading_dates
    tdates_fpath = '{0}{1}//TRADINGDATES.txt'.format(settings.data_path, settings.ticker)
    trading_dates = open(tdates_fpath, 'r').read().splitlines()

    import_option_data()

    print("Loading metadata complete.")

def import_option_data():
    global option_contracts
    global option_chains

    manifest_fpath = '{0}{1}//MANIFEST.csv'.format(settings.data_path, 
    settings.ticker)
    with open(manifest_fpath, mode='r') as manifest_file:

        csv_reader = csv.reader(manifest_file, delimiter=',')
        for row in csv_reader:
            
            chain_id = row[0]
            contract_id = row[1]
            option_symbol = row[2]
            expiration_date = row[3]
            option_type = row[4]
            strike = float(row[5])
            start_date = row[6]

            if not chain_id in option_chains:
                chain = option_chain.OptionChain(chain_id, option_symbol, expiration_date, start_date)
                option_chains[chain_id] = chain

            if not contract_id in option_contracts:
                contract = option_contract.OptionContract(contract_id, chain_id, strike, option_type, start_date)
                option_contracts[contract_id] = contract

            target_chain = option_chains[chain_id]

            if option_type == 'C':
                target_chain.call_options.append(option_contracts[contract_id])
            elif option_type == 'P':
                target_chain.put_options.append(option_contracts[contract_id])